update #555 - add development setup page (#632)
garageScript/c0d3-app: pages/docs/submitting.mdx

import LessonLayout from '../../components/mdx/Layout'

# Submitting Challenges

> Create an account on [c0d3.com/signup](http://c0d3.com/signup) if you don't

Reviewer: relative href

```suggestion
Create an account on [c0d3.com/signup](/signup) if you don't
```

**terminal**.

These steps are the same set of instructions you will most likely follow at
work. Each challenge you solve is simulating building a feature at work.

If this is all new to you, these steps may seem very intimidating. When you
start a new job, you will be equally intimidated because of all the specific
tools each company uses. It is important to ask questions when you feel
overwhelmed, there will be people on the chatroom help answer your questions.

## Local Setup

When you join a software engineering company, the first step is setting up your
machine, downloading the codebase, and setting up the codebase.
([video instructions](https://www.youtube.com/watch?v=JnRgR7z0rFU&t=254s))

1. After installing `node`, install c0d3 CLI (used to submit challenges):
   `npm i -g c0d3`

   - If you run into a `Missing write access ... EACCES` error, you need to run
     npm with sudo:
     - `sudo npm i -g c0d3`

1. Install [git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
   and familiarize yourself with [common git commands](/docs/git)
1. Use git to [clone the curriculum](https://github.com/garageScript/curriculum)
   into your machine

>    - `git clone [https://github.com/garageScript/curriculum](https://github.com/garageScript/curriculum)`

Reviewer: broken command

```suggestion
   - `git clone https://github.com/garageScript/curriculum`
```

   the c0d3 CLI tool used to submit the challenges, by typing `npm i -g c0d3` on
   the terminal at the bottom of the screen. **Note:** every time your workspace
   is restarted you'll need to run this command again.
3. After installing the CLI tool, type `c0d3 login` and enter your C0D3 account
   credentials to be able to submit the challenges.
4. Now the only thing remaining is to download the necessary libraries by
   running `yarn install`.

## Solving Challenges

`git status` is your best friend. It tells you what branch you are currently on
and which files are changed and/or staged for commit.

Solving each challenge is like building a feature at work. You must always
checkout `checkout master` before creating a new branch to work on a new
feature.

Branch names: Usually branch names are for your own eyes only, so you do not
have create branches with meaningful names.

_Note: Except for steps 3 (reading the challenge) and 4 (open and editing the
file), all the commands should be run in the terminal_

1. Make sure you are on the master branch: `git checkout master`
2. Create a new branch from master (pro tip: `p1` for problem 1)
   - `git checkout -b p1`

> 3. Go to [c0d3.com](http://c0d3.com), click on Curriculum, select lesson 0, and

Reviewer: link to relative path or to `https://www.c0d3.com` to skip the redirection server.

```suggestion
3. Go to [c0d3.com](/), click on Curriculum, select lesson 0, and
```

     it, and see the solution you just submitted! :)
8. Go back to **step 1** and repeat for the next challenge.

Make sure you go back to STEP 1 and run `git checkout master` so you are on
master branch before creating a new branch so you can start the next problem
fresh. If you don't then you will be creating a new branch from existing branch
and all your solutions will be piled up together.

**If you accidentally made a commit to your `master` branch, you need to undo
the commit.**

1. Undo the last commit: `git reset HEAD~`
2. Go to the branch for your problem (we will use `p1` as an example)
   - If the branch does not exist, you need to create one
     - `git checkout -b p1`
   - If the branch exists, you can just switch to it (by omitting the `-b`
     option)
     - `git checkout p1`
3. Continue with step 6 and onwards (add and commit)

## Getting Feedback

As a software engineer, after submitting a feature you will wait for other
software engineers to review your code.

Similarly, after submitting your solution, your solution will be reviewed by
other engineers to make sure you are following best practices in your code.

> You should be able to see your submissions when you go to
> [c0d3.com](http://c0d3.com), click on Curriculum, and select lesson 0.

Reviewer: link to curriculum here would be better, also some better wording.

```suggestion
You should be able to see your submissions when you go to the
[Curriculum](/curriculum) and selecting a lesson.
```
